File: Day99/Part3/temporal_skill_tree.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <set>

using namespace std;

// Problem 3: Skill Tree With Temporal Locks
// State: (Node, Time).
// BFS.
// Constraints: Time intervals.
// Inputs: N, T_target, StartTime, Edges (u, v, t_min, t_max, dt, cost).
// Cost (Energy) check. State: (Node, Time, Energy).
// Since Energy is resource, we want Min Energy for (Node, Time)?
// Or just reachability? "Energi terbatas" implies distinct limit or minimization.
// I'll assume standard MaxEnergy input and cost subtraction.

struct Edge {
    int v;
    int t_min, t_max, dt, cost;
};

int main() {
    int N, Target, StartTime, Energy;
    if (!(cin >> N >> Target >> StartTime >> Energy)) return 0;
    
    int M; cin >> M;
    vector<vector<Edge>> adj(N + 1);
    for(int i=0; i<M; ++i) {
        int u, v, tmin, tmax, dt, c;
        cin >> u >> v >> tmin >> tmax >> dt >> c;
        adj[u].push_back({v, tmin, tmax, dt, c});
    }
    
    // BFS State: u, time, energy
    // Visited optimization: Max energy at (u, time).
    // If we reach (u, time) with less energy, it's useless.
    // Time can range? Assume bounds? Or map?
    // Using map for sparse time.
    
    queue<tuple<int, int, int>> q;
    map<pair<int, int>, int> max_energy_at;
    
    q.push({1, StartTime, Energy}); // Assume Start Node 1
    max_energy_at[{1, StartTime}] = Energy;
    
    while(!q.empty()) {
        auto [u, t, e] = q.front(); q.pop();
        
        if (e < max_energy_at[{u, t}]) continue; // Outdated state
        
        if (u == Target) {
            cout << "YES" << endl;
            return 0;
        }
        
        for(const auto& edge : adj[u]) {
            if (t >= edge.t_min && t <= edge.t_max) {
                if (e >= edge.cost) {
                    int next_t = t + edge.dt;
                    int next_e = e - edge.cost;
                    
                    if (max_energy_at.find({edge.v, next_t}) == max_energy_at.end() || 
                        max_energy_at[{edge.v, next_t}] < next_e) {
                        
                        max_energy_at[{edge.v, next_t}] = next_e;
                        q.push({edge.v, next_t, next_e});
                    }
                }
            }
        }
    }
    
    cout << "NO" << endl;
    return 0;
}

File: Day99/Part3/temporal_skill_tree.py
import sys
from collections import deque, defaultdict

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    iterator = iter(input_data)
    try:
        N = int(next(iterator))
        Target = int(next(iterator))
        StartTime = int(next(iterator))
        Energy = int(next(iterator))
        M = int(next(iterator))
        
        adj = defaultdict(list)
        for _ in range(M):
            u = int(next(iterator))
            v = int(next(iterator))
            tmin = int(next(iterator))
            tmax = int(next(iterator))
            dt = int(next(iterator))
            c = int(next(iterator))
            adj[u].append((v, tmin, tmax, dt, c))
    except StopIteration:
        return

    # BFS state: (u, t) -> max_energy
    queue = deque([(1, StartTime, Energy)]) # Start Node 1
    max_energy_at = {} 
    max_energy_at[(1, StartTime)] = Energy
    
    while queue:
        u, t, e = queue.popleft()
        
        if e < max_energy_at.get((u, t), -1):
            continue
            
        if u == Target:
            print("YES")
            return
            
        for v, tmin, tmax, dt, cost in adj[u]:
            if tmin <= t <= tmax:
                if e >= cost:
                    nxt_t = t + dt
                    nxt_e = e - cost
                    
                    if max_energy_at.get((v, nxt_t), -1) < nxt_e:
                        max_energy_at[(v, nxt_t)] = nxt_e
                        queue.append((v, nxt_t, nxt_e))
                        
    print("NO")

if __name__ == "__main__":
    solve()
